skorch/utils.py

```diff
@@ -466,3 +467,19 @@ def get_map_location(target_device, fallback_device='cpu'):
             ), DeviceWarning)
         map_location = torch.device(fallback_device)
     return map_location
+
+
+class TeeGenerator:
+    """Stores a function that produces a generator and calls list on the
+    generator on the first iteration. All future iteration calls will
+    iterate over the list.
+    """
+    def __init__(self, gen):
+        self.gen = gen
+        self.gen_ = None
+
+    def __iter__(self):
+        if self.gen_ is None:
+            self.gen_ = self.gen()
```

Comment:
I think the TeeGenerator should use a generator, not a callable. The overhead is negligible but it allows for easy passing of partially consumed generators (e.g. when using models with two parameter sets) and is more intuitive.